### ISSUE TITLE
Fixing debconf/client helpers

### ISSUE THREAD

**Issue by julian-klode** [hughsie/PackageKit#257]
We currently have the problem that our client helper runs in the process using packagekit-glib2, meaning that if the process ends (for example, it crashes), no questions can be asked anymore; meaning that default values will be used and/or the installation will crash.

We thus need to run the client-helper as an independent process that is socket activated by systemd user sessions, and stops after some time out in order to not have it running all the time.

The steps I was thinking about are:

1. Fix pk-client-helper.c to remove a child that is done from the children array - currently we have a memory leak here - each package asking questions causes a new client helper child process to be created (a connection to be accepted), so we could keep a large number of children.

2. Add a `pk_client_helper_add_timeout(PKClientHelper *, int timeout, <some callback>)` that will add a new glib timeout function that checks whether the array of children is empty, and then calls the callback.

3. Add a `pk_client_helper_start_from_fd()` that is like `pk_client_helper_start()` but instead of a socket filename takes an open socket fd.

4. Add a `pk-debconf-helper.c` that uses the previous function with the fd passed by systemd and adds a timeout of 30 seconds. Install that as a user session service that is socket-activated, listening to `$XDG_RUNTIME_DIR/debconf.socket`.

5. Modify `pk_client_create_helper_socket()` in `pk-client.c` to return `frontend-socket=$XDG_RUNTIME_DIR/debconf.socket` if that socket exists, instead of generating a new in-process client helper.

I think 2-4 are easy, it's just 1 that I'm having a hard time with. Basically we have to set the socket and stdout channels to NULL when closing them, or we have to check in the function in 2 if they are closed.

**Comment by julian-klode** [comment]
Frequently raised concerns (FRC):

* Security: Currently we randomize the filename of the socket for each transaction. With the systemd-activation, we won't. Randomizing the filename does not really provide any protection, though - we could easily lookup any gpk-socket and replace it with our own in a malicious program. It's highly unlikely that debconf questions will have been asked by that time (and hence the socket is only listened to, but no connections to it have been made)

* UI integration: There was some concern from @ximion  that this leads to wrong window assignment in stuff. This is already the case, however - all windows belong to debconf-communicate.

**Comment by julian-klode** [comment]
Oh, for 1, I think we can check if the sources are still part of the main loop. If there are any children with valid sources left, we don't exit and wait for the next iteration of the idle loop.

**Comment by ximion** [comment]
Sounds reasonable to me :-)

**Comment by ximion** [comment]
No, this is done - there was a bit of a gap in maintenance at the time and I didn't get to clean up the older issues in the bugtracker yet :-)
